### PR TITLE
Add renamed vars to callee scope such that it is self consistent.

### LINKED ISSUE #7802: NumbaIRAssumptionWarning variable not in scope
Thanks to all the Numba devs. This is a great library! 😄 

I've been consistently getting `NumbaIRAssumptionWarning` in my library [galois](https://github.com/mhostetter/galois) since Numba 0.54.0. I previously opened #7623 and I believe this is related.

I've copied code out of my library into this minimal example. I'm calling a JIT function inside a ufunc. While both Numba 0.54.1 and 0.55.1 produce the correct result in *this* example, I've found times in my library where it is *not* producing the correct result. I fear the "variable not in scope" warnings are true and variables are being updated outside of the function scope.

**Q: What does `NumbaIRAssumptionWarning` mean in general?**
I'm trying to debug these warnings from my library and I'm having the hardest time.

**Q: What does `NumbaIRAssumptionWarning` mean in the Numba 0.55.1 case from my example?**
What's the difference between `_power_calculate_v6_b_3` and `_power_calculate_v6_b_4`? How do I know which variable (or one which line) it is referring?

For reference, issues I opened in my repo:
* https://github.com/mhostetter/galois/issues/178
* https://github.com/mhostetter/galois/issues/226

### Minimal example

```python
import numpy as np
import numba
print(np.__version__)
print(numba.__version__)


@numba.extending.register_jitable(inline="always")
def _multiply_calculate(a, b, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY):
    """
    a in GF(2^m), can be represented as a degree m-1 polynomial a(x) in GF(2)[x]
    b in GF(2^m), can be represented as a degree m-1 polynomial b(x) in GF(2)[x]
    p(x) in GF(2)[x] with degree m is the irreducible polynomial of GF(2^m)

    a * b = c
        = (a(x) * b(x)) % p(x) in GF(2)
        = c(x)
        = c
    """
    ORDER = CHARACTERISTIC**DEGREE

    # Re-order operands such that a > b so the while loop has less loops
    if b > a:
        a, b = b, a

    c = 0
    while b > 0:
        if b & 0b1:
            c ^= a  # Add a(x) to c(x)

        b >>= 1  # Divide b(x) by x
        a <<= 1  # Multiply a(x) by x
        if a >= ORDER:
            a ^= IRREDUCIBLE_POLY  # Compute a(x) % p(x)

    return c


MULTIPLY_JIT = numba.jit(["int64(int64, int64, int64, int64, int64)"], nopython=True, cache=True)(_multiply_calculate)


@numba.extending.register_jitable(inline="always")
def _power_calculate(a, b, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY):
    """
    Square and Multiply Algorithm

    a^13 = (1) * (a)^13
            = (a) * (a)^12
            = (a) * (a^2)^6
            = (a) * (a^4)^3
            = (a * a^4) * (a^4)^2
            = (a * a^4) * (a^8)
            = result_m * result_s
    """
    if a == 0 and b < 0:
        raise ZeroDivisionError("Cannot compute the multiplicative inverse of 0 in a Galois field.")

    if b == 0:
        return 1
    # elif b < 0:
    #     a = RECIPROCAL(a, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY)
    #     b = abs(b)

    result_s = a  # The "squaring" part
    result_m = 1  # The "multiplicative" part

    while b > 1:
        if b % 2 == 0:
            result_s = MULTIPLY_JIT(result_s, result_s, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY)
            b //= 2
        else:
            result_m = MULTIPLY_JIT(result_m, result_s, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY)
            b -= 1

    result = MULTIPLY_JIT(result_m, result_s, CHARACTERISTIC, DEGREE, IRREDUCIBLE_POLY)

    return result


characteristic = 2
degree = 2
irreducible_poly = 7

POWER_UFUNC = numba.vectorize(["int64(int64, int64)"], nopython=True)(lambda aa, bb: _power_calculate(aa, bb, characteristic, degree, irreducible_poly))

x = np.array([0, 1, 2, 3])  # Elements
y = np.array([0, 1, 2, 3])  # Powers

z = POWER_UFUNC.outer(x, y)
print(z)

z_truth = np.array([[1, 0, 0, 0], [1, 1, 1, 1], [1, 2, 3, 1], [1, 3, 2, 1]])
print(z_truth)

print(np.array_equal(z, z_truth))
```

### Numba 0.54.1

```
1.20.3
0.54.1
[[1 0 0 0]
 [1 1 1 1]
 [1 2 3 1]
 [1 3 2 1]]
[[1 0 0 0]
 [1 1 1 1]
 [1 2 3 1]
 [1 3 2 1]]
True
```

### Numba 0.55.1

```
1.20.3
0.55.1
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_result_s_3' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_b_4' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_result_m_3' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_result_s_2' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_b_3' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
/home/matt/.local/lib/python3.8/site-packages/numba/core/ssa.py:272: NumbaIRAssumptionWarning: variable '_power_calculate_v6_result_m_2' is not in scope.

This warning came from an internal pedantic check. Please report the warning
message and traceback, along with a minimal reproducer at:
https://github.com/numba/numba/issues/new?template=bug_report.md

  warnings.warn(errors.NumbaIRAssumptionWarning(wmsg,
[[1 0 0 0]
 [1 1 1 1]
 [1 2 3 1]
 [1 3 2 1]]
[[1 0 0 0]
 [1 1 1 1]
 [1 2 3 1]
 [1 3 2 1]]
True
```

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Fixes #7802

### CODE REVIEW

**Comment by mhostetter** [comment]
@stuartarchibald I'm trying to test your branch with my "failure" condition from #7802. I'm having difficulty installing your branch, though.

I did this:

```
$ pip3 install git+https://github.com/stuartarchibald/numba.git@fix/7802
Collecting git+https://github.com/stuartarchibald/numba.git@fix/7802
  Cloning https://github.com/stuartarchibald/numba.git (to revision fix/7802) to /tmp/pip-req-build-_c6kebyr
  Running command git clone -q https://github.com/stuartarchibald/numba.git /tmp/pip-req-build-_c6kebyr
  Running command git checkout -b fix/7802 --track origin/fix/7802
  Switched to a new branch 'fix/7802'
  Branch 'fix/7802' set up to track remote branch 'fix/7802' from 'origin'.
ERROR: Could not find a version that satisfies the requirement llvmlite<0.40,>=0.39.0dev0 (from numba==0.50.0.dev0+4742.ge2400df9e) (from versions: 0.2.0, 0.2.1, 0.2.2, 0.4.0, 0.5.0, 0.6.0, 0.7.0, 0.8.0, 0.9.0, 0.10.0, 0.11.0, 0.12.0.1, 0.12.1, 0.13.0, 0.14.0, 0.15.0, 0.16.0, 0.17.0, 0.17.1, 0.18.0, 0.19.0, 0.20.0, 0.21.0, 0.22.0, 0.23.0, 0.23.2, 0.24.0, 0.25.0, 0.26.0, 0.27.0, 0.27.1, 0.28.0, 0.29.0, 0.30.0, 0.31.0, 0.32.0, 0.32.1, 0.33.0rc1, 0.33.0, 0.34.0rc1, 0.34.0, 0.35.0rc2, 0.35.0rc3, 0.35.0, 0.36.0rc1, 0.36.0rc2, 0.36.0, 0.37.0rc2, 0.37.0, 0.38.0rc1, 0.38.0)
ERROR: No matching distribution found for llvmlite<0.40,>=0.39.0dev0 (from numba==0.50.0.dev0+4742.ge2400df9e)
```

Should I have done something different?

**Comment by stuartarchibald** [comment]
@mhostetter Thanks for taking a look, the `main` branch of Numba generally needs development versions of `llvmlite`, these are available from the `numba` conda channel under the `dev` label.

**Review by sklam — APPROVED** [review]
Confirmed that this patch removes the scope warning

File: numba/core/inline_closurecall.py [hunk 1]
```diff
@@ -407,11 +407,17 @@ def copy_ir(the_ir):
...
+                # Update the caller scope with the new names
                 new_var = scope.redefine(inlined_name, loc=var.loc)
+                # Also update the callee scope with the new names. Should the
+                # type and call maps need updating (which requires SSA form) the
+                # transformation to SSA is valid as the IR object is internally
+                # consistent.
+                callee_scope.redefine(inlined_name, loc=var.loc)
                 var_dict[var.name] = new_var
         self.debug_print("var_dict = ", var_dict)
         replace_vars(callee_blocks, var_dict)
```
Comment (sklam) [inline numba/core/inline_closurecall.py:423]: The same effect can be achieved by:

```suggestion
                # Also update the callee scope with the new names. Should the
                # type and call maps need updating (which requires SSA form) the
                # transformation to SSA is valid as the IR object is internally
                # consistent.
                # callee_scope.redefine(inlined_name, loc=var.loc)
                var_dict[var.name] = new_var
        self.debug_print("var_dict = ", var_dict)
        replace_vars(callee_blocks, var_dict)

        from numba.core.ir_utils import fixup_var_define_in_scope
        fixup_var_define_in_scope(callee_blocks)
```

Perhaps, `fixup_var_define_in_scope` should be part of `replace_vars`

**Pull request merged**